### PR TITLE
Use a trimmed version of README.md for PyPI

### LINKED ISSUE #33505: Trim down the PyPI readme
### Body

The PyPI README is far too long for the PyPI page.  Just basic summary is needed and links to documentation, changelog,GitHub page.

We have now dedicated Changelog page and there should be a "project-url" pointing to it (same as in Providers). 

Some images (AWS) are broken and they should not be there anyway.

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by pankajkoti**
closes: #33505 

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
I think we do NOT want to maintain yet another README. My vision of is that the README will py automatically extracted from the main README by pre-commit (so ideally it should be placed in `generated` dir and used from there, 

We could do it dynamically as well, but that would not fly with mostly declarative way how we want to build packages. So automatically generated file whic is just truncated version of README.md commited in `generated`  seems a much better approach.

**Comment by pankajkoti** [comment]
> I think we do NOT want to maintain yet another README. My vision of is that the README will be automatically extracted from the main README by pre-commit (so ideally it should be placed in `generated` dir and used from there,
> 
> We could do it dynamically as well, but that would not fly with mostly declarative way how we want to build packages. So automatically generated file which is just truncated version of README.md commited in `generated` seems a much better approach.

yes, good idea, will rather work on the extraction via pre-commit.

**Review by potiuk — APPROVED** [review]
love it

**Comment by potiuk** [comment]
Really tiny NIT. in `generated/README.md` I explained how various files in that folder are generated. Would be great to add a paragrapgh descreiption on where it is from. When you look at that file, it's not obvious which commit generates it and how, so adding a short explanation will aid future contributors (and even future selves) to know where the hell that fille is from

**Comment by potiuk** [comment]
Super - Super NIT. It would also be great if the first line of the generated PYPIREADME.md contains something like that
```
<!-- PLEASE DO NOT MODIFY THIS FILE. IT HAS BEEN GENERATED AUTOMATICALY FROM THE `README.md` IN MAIN AIRFLOW REPO BY `....` PRE-COMMIT. YOUR CHANGES HERE WILL BE AUTOMATICALLY OVERWRITTEN -->  
```

You can also safely remove the Licence header when you generate it (though this might be a bit complex so we can leave it). The ASF requirement is that licence headers are present in files that are human-prepared, it does not have to be added to files which are automatically generated from other files.

**Comment by potiuk** [comment]
BTW. I REALLY wanted to add such comment in the .json files - but unfortunately, json specification does not allow comments.

File: scripts/ci/pre_commit/pre_commit_generate_pypi_readme.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    else:
+        raise Exception(f"Cannot find section {section_name} in README.md")
+
+
+if __name__ == "__main__":
+    readme_file = AIRFLOW_SOURCES / "README.md"
+    pypi_readme_file = AIRFLOW_SOURCES / "generated" / "PYPI_README.md"
+
+    if not pypi_readme_file.exists():
+        pypi_readme_content = PYPI_README_HEADER
```
Comment (pankajkoti) [inline scripts/ci/pre_commit/pre_commit_generate_pypi_readme.py]:
```suggestion
        pypi_readme_content = ""
```

File: setup.cfg [hunk 2]
```diff
@@ -22,7 +22,7 @@ author = Apache Software Foundation
 author_email = dev@airflow.apache.org
 url = https://airflow.apache.org/
 version = attr: airflow.__version__
-long_description = file: README.md
+long_description = file: generated/PYPI_README.md
```
Comment (pankajkoti) [inline setup.cfg:25]: The generated folder is included in the `MANIFEST.in `https://github.com/apache/airflow/blob/main/MANIFEST.in#L42

So I am guessing this file will be available correctly and be used for the PyPI project listing description?
If someone can confirm my guess, would be great.

Comment (uranusjr) [reply setup.cfg:25]: You can build the package with `breeze release-management prepare-airflow-package --package-format sdist` and extract it to check if it’s in the package.

Comment (potiuk) [reply setup.cfg:25]: Yes. And use "both" to generate both packages and you can also check if it is in METADATA of the .whl package

**Review by potiuk — APPROVED** [review]
Looks great now.

File: setup.cfg (hunk 2 above)
Comment (pankajkoti) [reply setup.cfg:25]: Thank you.

Yes, confirmed that the file is included both in the `.tar.gz` and in the wheel's pkg description

Screenshot 2023-08-30 at 2 44 28 PM

**Review by uranusjr — APPROVED** [review]

**Comment by pankajkoti** [comment]
Static check is failing with the below error:
```
Check if licenses are OK for Apache................................................Failed
- hook id: check-apache-license-rat
- exit code: 1

Running command:
docker run -v /home/runner/work/airflow/airflow:/opt/airflow -t --user 1001:999 --rm --platform linux/amd64 ghcr.io/apache/airflow-apache-rat:0.13-2021.07.04 -d /opt/airflow --exclude-file /opt/airflow/.rat-excludes

ERROR: Could not find Apache licences in some files:

 !????? /opt/airflow/generated/PYPI_README.md
 ```
 
 I tried adding `generated/PYPI_README.md` (this is already committed and fails in this PR checks), `./generated/PYPI_README.md`, `/opt/airflow/generated/PYPI_README.md ` to the `.rat-excludes` file, also tried to rebuild the image, but the error persists when I run `breeze static-checks --type check-apache-license-rat --all-files --force-build` locally.
 
 Am I missing something obvious or there's more to do 🤔

File: setup.cfg [hunk 3]
```diff
@@ -43,11 +43,12 @@ classifiers =
     Topic :: System :: Monitoring
     Framework :: Apache Airflow
 project_urls =
+    Bug Tracker=https://github.com/apache/airflow/issues
+    Changelog=https://airflow.apache.org/docs/apache-airflow/stable/release_notes.html
```
Comment (jedcunningham) [inline setup.cfg]:
```suggestion
    Release Notes=https://airflow.apache.org/docs/apache-airflow/stable/release_notes.html
```

And maybe reorder it as well...

File: scripts/ci/pre_commit/pre_commit_generate_pypi_readme.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    for section in README_SECTIONS_TO_EXTRACT:
+        section_content = extract_section(readme_content, section)
+        generated_pypi_readme_content += section_content
+    if pypi_readme_content != generated_pypi_readme_content:
+        with pypi_readme_file.open("w") as generated_readme:
+            generated_readme.write(generated_pypi_readme_content)
+        console.print("\n[yellow]PyPI README.md content regenerated after update..\n")
+        console.print("[red]Aborting the commit")
+        console.print("You should check the changes made. Then simply 'git add --update .' and re-commit")
+        sys.exit(1)
```
Comment (jedcunningham) [inline scripts/ci/pre_commit/pre_commit_generate_pypi_readme.py]: How about we just write the file every time? That simplifies this code a bit, and still shows a good enough message to the user:

```
Generate PyPI README...............................................................Failed
- hook id: generate-pypi-readme
- files were modified by this hook
```

```suggestion
    readme_content = readme_file.read_text()

    generated_pypi_readme_content = PYPI_README_HEADER
    for section in README_SECTIONS_TO_EXTRACT:
        section_content = extract_section(readme_content, section)
        generated_pypi_readme_content += section_content

    with pypi_readme_file.open("w") as generated_readme:
        generated_readme.write(generated_pypi_readme_content)
```

(If not, I have a number of comments on the original code, so lmk 👍)

Comment (pankajkoti) [reply scripts/ci/pre_commit/pre_commit_generate_pypi_readme.py]: yes, this works very well. Very clean, thank you :) 

Yes, I had all the ugly code earlier and the rich dependency for the console output & failing the commit with sys.exit which is no longer needed 🚀

Comment (potiuk) [reply scripts/ci/pre_commit/pre_commit_generate_pypi_readme.py]: Yeah. Writing file every time is fine.  Now. The only thing that's left is just to ... run the pre-commit to fix some discrepancies :)

Comment (pankajkoti) [reply scripts/ci/pre_commit/pre_commit_generate_pypi_readme.py]: I am still seeing static failures wrt https://github.com/apache/airflow/pull/33637#issuecomment-1699142821. Any suggestion for me that what we could do here?

**Comment by potiuk** [comment]
> Am I missing something obvious or there's more to do 🤔

Yep. I think you found the fix in the last commit :)

**Comment by pankajkoti** [comment]
> Yep. I think you found the fix in the last commit :)

I am not sure. I did add the file to .rat-excludes, but it still fails, waiting for another run in the CI now, but I'm not hopeful.

File: .rat-excludes [hunk 5]
```diff
@@ -132,3 +132,6 @@ chart/*.iml
 
 # Sha files
 .*sha256
+
+# Generated PyPI README file
+generated/PYPI_README.md
```
Comment (potiuk) [inline .rat-excludes]:
```suggestion
./generated/*
```

**Comment by alexbegg** [comment]
> BTW. I REALLY wanted to add such comment in the .json files - but unfortunately, json specification does not allow comments (hence the README.md).

@potiuk maybe it is it possible to switch these files to .jsonc, "JSON with Comments"? VSCode supports it for example (https://code.visualstudio.com/docs/languages/json#_json-with-comments), I am not sure if it is possible to easily include support in this project.

**Comment by potiuk** [comment]
> > BTW. I REALLY wanted to add such comment in the .json files - but unfortunately, json specification does not allow comments (hence the README.md).
> 
> @potiuk maybe it is it possible to switch these files to .jsonc, "JSON with Comments"? VSCode supports it for example (https://code.visualstudio.com/docs/languages/json#_json-with-comments), I am not sure if it is possible to easily include support in this project.

Well. We need  to read them in setup.py using nothing by stdlib of python in all versions we support. Otherwise I'd just use YAML. The file is read by `python setup.py` that should work without being in any specifc venv, no libraries installed, no IDEs or anything. Nothing beats simplicity of `import json; provider_dependencies_dict =json.loads( Path(__name__.parent) / "generated" /"provider_dependencies.json").text)`

**Comment by Taragolis** [comment]
> Well. We need to read them in setup.py using nothing by stdlib of python in all versions we support. Otherwise I'd just use YAML.

We just need wait a bit (3 years 😀 ) for min Python 3.11 and we could use tomllib

**Comment by pankajkoti** [comment]
I added the license back to the generated file as the the exclude is not working. 
Static checks seems to run fine now.

One failing test of [MSSQL2017-latest, Py3.8: Core Providers[-amazon,google] Other Providers[amazon] WWW API Always CLI Providers[google]](https://github.com/apache/airflow/actions/runs/6046265544/job/16408363681?pr=33637#logs) seems to be unrelated.

**Comment by phanikumv** [comment]
> I added the license back to the generated file as the the exclude is not working. Static checks seems to run fine now.
> 
> One failing test of [MSSQL2017-latest, Py3.8: Core Providers[-amazon,google] Other Providers[amazon] WWW API Always CLI Providers[google]](https://github.com/apache/airflow/actions/runs/6046265544/job/16408363681?pr=33637#logs) seems to be unrelated.

just re-triggered the failed test, lets see if its a flaky one

**Review by potiuk — APPROVED** [review]

**Pull request merged**